amarna/rules/post_process_rules/StorageVarCollisionRule.py

```diff
@@ -0,0 +1,43 @@
+from typing import Dict, List
+from amarna.Result import ResultMultiplePositions, result_multiple_positions
+
+from amarna.rules.gatherer_rules.DeclaredFunctionsGatherer import (
+    DeclaredFunctionsGatherer,
+    FunctionType,
+)
+
+class StorageVarCollisionRule:
+    """
+    Storage variable is declared multiple times across files.
+    """
+
+    RULE_TEXT = "[This](0) storage variable is also declared [here](1)."
+    RULE_NAME = "storage-var-collision"
+
+    def run_rule(self, gathered_data: Dict) -> List[ResultMultiplePositions]:
+        declared_functions: List[FunctionType] = gathered_data[
+            DeclaredFunctionsGatherer.GATHERER_NAME
+        ]
+
+        results = []
+
+        storage_vars = filter(lambda func: "storage_var" in func.decorators, declared_functions)
```

Comment:
`filter` returns an iterator in python. Every time you will access the iterator, you will "consume" one of its element.

Because here the filter is accessed in nested loops (L26/L27), after the first iteration, all the elements of the iterator will be consumed. 

As a result the for/for loops structure here lead to compare only the first element in the storage vars for duplicate.

To see this behavior, you can run
```python
l = [1,2,3,4,6,6]


my_filter = filter(lambda x: x>3, l)

for elem in my_filter:
  for other_elem in my_filter:
    print(f"compare {elem} with {other_elem}")
```

Which will lead to 
```
compare 4 with 6
compare 4 with 6
```
Here the two 6 are never compared against each other.

To prevent that you can do `list(filter(...`, so that the iterator is converted to a list. However I think by doing so, you will end up having duplicate results (e.g. a.balance being compared against b.balance, and then b.balance being compared against a.balance)

Instead of using a double for, you can use different structure. For example, you can first create a dict that would map the name to a list of FunctionType, and then report any element of the dict if it has more than 1 element in the list. [itertools](https://docs.python.org/3/library/itertools.html) might also be useful here (but it might be overkill)

Using the dict approach will also help to group together the duplicate if there are more than 2 variables with the same name.

Reply:
Good catch, that explains a bug I was having actually. could you have a look now? it creates a dictionary as you said, mapping the names to a list of FunctionType

Reply:
Good catch, that explains a bug I was having actually. could you have a look now? it creates a dictionary as you said, mapping the names to a list of FunctionType

Reply:
Something else I'd like your opinion on is whether warning on every duplicate storage_var is the right approach here. Is it always an issue? Or only if the two files with the same storage_var are imported somewhere simultaneously?